org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/core/IRepl.java

```diff
@@ -0,0 +1,44 @@
+package org.metaborg.spoofax.shell.core;
+
+import org.metaborg.core.MetaborgException;
+import org.metaborg.spoofax.shell.commands.IReplCommand;
+import org.metaborg.spoofax.shell.invoker.CommandNotFoundException;
+import org.metaborg.spoofax.shell.invoker.ICommandInvoker;
+
+/**
+ * This interface defines the evaluation part of a REPL (Read-Eval-Print-Loop). The reason for only
+ * defining the evaluation phase is the fact that different clients have different needs and as
+ * such, there is no way of defining methods {@code read} or {@code print} methods suitable for all
+ * possible clients.
+ *
+ * The default {@link #eval(String)} implementation uses an {@link ICommandInvoker} to process user
+ * input. This {@link ICommandInvoker} invokes {@link IReplCommand}s, which use
+ * {@link org.metaborg.spoofax.shell.hooks} to display output.
+ */
+public interface IRepl {
```

Comment:
Skip mentioned here that the only behavior to implement for a client is the `getInvoker` getter, which kind of defeats the purpose of an interface. I agree, but according to Gerlof having an abstract class at the top of your hierarchy is also not "nice", so I'm not sure which solution you prefer here.